controlplane/internal/controller/clusterdeployment_controller.go

```diff
@@ -273,3 +285,92 @@ func (r *ClusterDeploymentReconciler) createImageRegistry(ctx context.Context, r
+// For MNO (multi-node openshift), this is set to the default list for baremetal platform,
+// but the OpenshiftAssistedControlPlane can specify additional capabilities to be appended to this list.
+func setACICapabilities(oacp *controlplanev1alpha2.OpenshiftAssistedControlPlane, aci *hiveext.AgentClusterInstall) error {
+	isBaremetalPlatform := aci.Spec.PlatformType == hiveext.BareMetalPlatformType
+	if !capabilitiesSet(oacp.Spec.Config.Capabilities) && !isBaremetalPlatform {
+		return nil
+	}
+	var installCfgOverride InstallConfigOverride
+	baselineCapability, err := getBaselineCapability(oacp.Spec.Config.Capabilities.BaselineCapability, isBaremetalPlatform)
+	if err != nil {
+		return err
+	}
+	installCfgOverride.Capability.BaselineCapability = baselineCapability
+
+	additionalEnabledCapabilities := getAdditionalCapabilities(oacp.Spec.Config.Capabilities.AdditionalEnabledCapabilities, isBaremetalPlatform)
+	installCfgOverride.Capability.AdditionalEnabledCapabilities = additionalEnabledCapabilities
+
+	installCfgOverrideStr, err := json.Marshal(installCfgOverride)
+	if err != nil {
+		return err
+	}
+
+	if aci.Annotations == nil {
+		aci.Annotations = make(map[string]string)
+	}
+
+	aci.Annotations[InstallConfigOverrides] = string(installCfgOverrideStr)
+	return nil
+}
+
+func getBaselineCapability(capability string, isBaremetalPlatform bool) (string, error) {
+	baselineCapability := capability
+	if baselineCapability == "None" || baselineCapability == "vCurrent" {
+		return baselineCapability, nil
+	}
+
+	if baselineCapability == "" {
+		baselineCapability = defaultBaselineCapability
+		if isBaremetalPlatform {
+			baselineCapability = defaultBaremetalBaselineCapability
+		}
+		return baselineCapability, nil
+	}
+
+	var baselineCapabilityRegexp = regexp.MustCompile(`v4\.[0-9]+`)
+	if !baselineCapabilityRegexp.MatchString(baselineCapability) {
+		return "",
+			fmt.Errorf("invalid baseline capability set, must be one of: None, vCurrent, or v4.x. Got: [%s]", baselineCapability)
+	}
+	return baselineCapability, nil
+}
+
+func getAdditionalCapabilities(specifiedAdditionalCapabilities []string, isBaremetalPlatform bool) []string {
+	additionalCapabilitiesList := []string{}
+	if isBaremetalPlatform {
+		additionalCapabilitiesList = append([]string{}, defaultBaremetalAdditionalCapabilities...)
+	}
+
+	for _, capability := range specifiedAdditionalCapabilities {
+		// Ignore MAPI for baremetal MNO clusters and ignore duplicates
+		if (strings.EqualFold(capability, "MachineAPI") && isBaremetalPlatform) || funk.ContainsString(additionalCapabilitiesList, capability) {
+			continue
+		}
+		additionalCapabilitiesList = append(additionalCapabilitiesList, capability)
+	}
+
+	if len(additionalCapabilitiesList) < 1 {
+		return nil
+	}
+
+	return additionalCapabilitiesList
+}
+
+func capabilitiesSet(capabilities v1alpha2.Capabilities) bool {
+	return capabilities.BaselineCapability != "" || len(capabilities.AdditionalEnabledCapabilities) > 0
```

Comment:
Wouldn't comparing zero value for Capabilities also work?

Reply:
<img width="587" alt="Screenshot 2025-03-18 at 10 53 45 AM" src="https://github.com/user-attachments/assets/29f89fee-97a2-4cc7-a070-88e234112d9f" />

Hmm I get this when I try to compare the empty struct... Do you know of another way to compare the zero value?

Reply:
Oh, right, I didn't realize, sorry :sweat: 

In any case I think we can solve it with apimachinery equality package. which we already use in the project:
```equality.Semantic.DeepEqual(myCapabilities, zeroCapabilities)```

Reply:
Oh no worries! Thank you for recommendation, I modified it to use the equality package